liana/method/sp/_misty.py

```diff
@@ -0,0 +1,471 @@
+            # TODO: make this faster/better
+            distance_weights_csr = distance_weights.copy()
+            distance_weights_csr.data[np.isin(distance_weights_csr.indices, np.where(adata.obs[group_env_by]!=group)[0])] = 0
+            paraviews[group] = ad.AnnData(X=distance_weights_csr@adata[:, predictors].X, obs=adata.obs, var=pd.DataFrame(index=predictors))
+    else:
+        paraviews["all"] = ad.AnnData(X=distance_weights@adata[:, predictors].X, obs=adata.obs, var=pd.DataFrame(index=predictors))
+    return(paraviews)
+
+
+def _get_juxtaview_groups(adata, predictors, group_env_by, juxta_cutoff=np.inf, add_self=True, spatial_key="spatial"):
+    neighbors = _get_neighbors(adata, juxta_cutoff=juxta_cutoff, add_self=add_self, spatial_key=spatial_key)
+    juxtaviews = {}
+    if group_env_by:
+        groups = np.unique(adata.obs[group_env_by])
+        for group in groups:
+            # see comment in get_paraview_groups
+            neighbors_csr = neighbors.copy()
+            neighbors_csr.data[np.isin(neighbors_csr.indices, np.where(adata.obs[group_env_by]!=group)[0])] = 0
+            juxtaviews[group] = ad.AnnData(X=neighbors_csr@adata[:, predictors].X, obs=adata.obs, var=pd.DataFrame(index=predictors))
+    else:
+        juxtaviews["all"] = ad.AnnData(X=neighbors@adata[:, predictors].X, obs=adata.obs, var=pd.DataFrame(index=predictors))
+    return(juxtaviews)
+
+
+def _compose_views_groups(xdata, predictors, bypass_intra, add_juxta, add_para, group_env_by, juxta_cutoff, bandwidth, kernel, zoi, add_self, spatial_key):
+    views = {}
+    if not bypass_intra:
+        views["intra"] = xdata
+    if add_juxta:
+        views["juxta"] = _get_juxtaview_groups(xdata, predictors, group_env_by=group_env_by, juxta_cutoff=juxta_cutoff, 
+                                               add_self=add_self, spatial_key=spatial_key)
+    if add_para:
+        views["para"] = _get_paraview_groups(xdata, predictors, bandwidth=bandwidth, group_env_by=group_env_by, kernel=kernel, 
+                                             add_self=add_self, spatial_key=spatial_key, zoi=zoi)
+    return(views)
+
+
+def _check_anndata_objects_groups(xdata, ydata, spatial_key, group_intra_by, group_env_by):
+    # check that the adata_source and adata_target have the same number of spots/cells
+    assert xdata.X.shape[0] == ydata.X.shape[0]
+
+    # check that the adata_source and adata_target have the same spatial coordinates
+    assert np.all(xdata.obsm[spatial_key] == ydata.obsm[spatial_key])
+    
+    # check the group_target_by and group_views_by variables
+    if group_intra_by is not None:
+        assert group_intra_by in ydata.obs.columns
+    if group_env_by is not None:
+        assert group_env_by in xdata.obs.columns
+
+
+def _check_target_in_predictors(target, predictors):
+    # TODO: this is only necessary for the intra view, for the other views we could make it optional
```

Comment:
to clarify with @schae211

Reply:
add `self_interactions` param with False as default - show example where self_interactions are of interest - e.g. cell types in labelled high res spatial

Reply:
This feature is implemented now, see parameter 'keep_same_predictor' (feel free to rename it)